proposals/NNNN-resource-arrays.md

```diff
@@ -0,0 +1,356 @@
+allocate such variable? The fact that DXC allows this should be considered a bug.
+
+Additionally, and as in the previous example, DXC does not report an error when an
+uninitialized resource in an unbounded array is accessed and instead creates a
+call to `dx.op.annotateHandle` with zero-initialized handle value.
+
+```
+RWBuffer<float> Buf;
+RWBuffer<float> Out;
+
+float foo() {
+  RWBuffer<float> Unbounded[];
+  Unbounded[100] = Buf;
+  // Unbounded[0] is not initialized
+  return Unbounded[100][0] + Unbounded[0][0];
+}
+
+[numthreads(4,1,1)]
+void main() {
+  Out[0] = foo();
+}
+```
+https://godbolt.org/z/43rGnY4Ee
+
+#### Resource Arrays as Function Arguments
+
+Both fixed-size and unbounded resource arrays can be used as function arguments.
+For fixed-size arrays, the dimensions of the array passed into a function must
+exactly match the dimensions of the declared argument. For unbounded arrays this
+restriction does not exist; a fixed-size array can be used as an argument of a
+function that accepts an unbounded array.
+
+```
+RWBuffer<float> K[3] : register(u0);
+RWBuffer<float> L[2][2] : register(u10);
+RWBuffer<float> M[] : register(u0, space1);
+RWBuffer<float> Out;
+
+float foo(RWBuffer<float> LK[3], RWBuffer<float> LL[2][2],
+          RWBuffer<float> LM1[], RWBuffer<float> LM2[]) {
+  return LK[2][0] + LL[1][1][0] + LM1[100][0] + LM2[100][0];
+}
+
+[numthreads(4,1,1)]
+void main() {
+  Out[0] = foo(K, L, M, K);
+}
+```
+https://godbolt.org/z/6Eh3fnME9
+
+Note that array `K` of size `3` is passed into function `foo` via an unbounded
+array argument and it is indexed beyond its size.
+
+#### Subsets of Multi-Dimensional Arrays
+ 
+Multi-dimensional array can be indexed to refer to a lower-dimensional subset of
+the array. For example, for `RWBuffer<float> N[10][5];` the expression `N[7]`
+refers to a sub-array in `N` of size `5`.
+
+```
+RWBuffer<float> N[10][5] : register(u0);
+RWBuffer<float> Out;
+
+float foo(RWBuffer<float> P[5]) {
+  return P[3][0];
+}
+
+[numthreads(4,1,1)]
+void main() {
+  Out[0] = foo(N[7]);
+}
+```
+
+https://godbolt.org/z/5cG8xWWa6
+
+Note that while the function argument `RWBuffer<float> P[5]` is a local array of
+size `5`, it actually refers to a subset of a larger multi-dimensional array
+`N`. The handle initialization call for resources in `P` must contain
+information about the original array's range.
+
+For example, the handle initialization for the resource that is referenced by `P[3]`
+will have lower bound `0`, upper bound `49`, and index `38 = 0 + 7*5 + 3`:
+
+```
+call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 0, i32 49, i32 0, i8 1 }, i32 38, i1 false)
+```
+https://godbolt.org/z/YejsdsTKc
+
+At the same time, `P` is a local variable and as such it is editable. It should
+be initialized by copy-in array semantics to refer to resource handles from the
+original larger array `N`, but its individual array elements can be overridden,
```

Comment:
individual array elements refers to the resource handles? Or the floats in the resource the resource handle points to?